File: leetcode/662_Maximum_Width_of_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        if (root == null) 
            return 0;
        
        int max = 1;
        int level = 0;
        Queue<TreeNode> queue = new LinkedList<>();
        Queue<Integer> queue2 = new LinkedList<>(); // for idx 
        queue.offer(root);
        queue2.offer(0);
        while (!queue.isEmpty()) {
            int left = Integer.MAX_VALUE;
            int right = -1;
            level++;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                Integer idx = queue2.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                    queue2.offer(idx * 2);
                    left = Math.min(idx * 2, left);
                    right = Math.max(idx * 2, right);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                    queue2.offer(idx * 2 + 1);
                    left = Math.min(idx * 2 + 1, left);
                    right = Math.max(idx * 2 + 1, right);
                }
            }
            if (right != -1 && left != Integer.MAX_VALUE) 
                max = Math.max(max, right - left + 1 );
        }
        return max;
    }
}

File: leetcode/662_Maximum_Width_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def dfs(self, r, dep, no):
        if r is None:
            return
        if dep not in self.d:
            self.d[dep] = []
        self.d[dep].append(no)
        self.dfs(r.left, dep + 1, no * 2)
        self.dfs(r.right, dep + 1, no * 2 + 1)
    def widthOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.d = dict()
        self.dfs(root, 0, 1)
        ans = 0
        for x in self.d:
            ans = max(ans, self.d[x][-1] - self.d[x][0] + 1)
        return ans